specs/carousel/carousel.md

```diff
@@ -0,0 +1,258 @@
+- `autoplay-interval` - a number representing the autoplay interval in milliseconds
+- `loop` - boolean value to toggle if the carousel loops slides at the beginning or end (defaults to true)
+- `paused` - whether the rotation is paused or not
+- `aria-roledescription` - set to *carousel* for overall component, and set to *slide* for each slide
+- `aria-labelledby` - optional based on implementation**
+- `aria-label` - optional, based on implementation**
+
+** See the [W3C Specification](https://w3c.github.io/aria-practices/#wai-aria-roles-states-and-properties-4) for requirements and details.
+
+*CSS custom properties:*
+- ???
+
+*Slots:*
+- `default` - slot for slides content // TODO: IS this the right way?
+- `play-toggle` - the control to toggle if the slides rotate or not
+- `next-flipper` - the control used to rotate to the next slide
+- `previous-flipper` - the control used to rotate to the previous slide
+- `tablist` - the control to select a certain slide
+
+*Events*
+- `on-pause` - callback fired when the rotation is stopped, such as when keyboard focus or hover is on any element of the carousel, or the *play-toggle* is set to off.
+- `on-active-id-update` - callback fired when the active id is changed
+
+### Anatomy and Appearance
+**Carousel Structure:**
+
+Note: the Carousel class will have a method that checks the attribute names and grabs the items.
+
+```html
+<template>
+    <div
+        class="carousel"
+        aria-roledescription="carousel"
+        tabindex="-1"
+    >
+    <slot name="play-toggle"> SEE DEFAULT </slot>
+    <slot name="next-flipper"> SEE DEFAULT </slot>
+    <div
+        class="carousel-items"
+        aria-live="off"
+    >
+        <!-- Solution 1 -->
+        <!-- Map all items and use css class's to display correct item based on id. -->
+        <!-- May be a less performant way as the items will be mapped each time. -->
+        ${ x => x.items.map(item, index => {
+            return `
+            <div
+                class="carousel-item ${x => x.getActiveIndex() === index ? 'active' : ''}"
+                role="group"
+                aria-roledescription="slide"
+                aria-label="${x => x.getActiveIndex()} of ${x => x.getItemCount()}"
+            >
+                ${item}
+            </div>`
+            }).join('')
+        }
+        <!-- Solution 2 -->
+        <!-- Have carousel class method that query selects this div and inserts content for the correct item based on id/index from the id's given as an array. -->
+        <div
+            class="carousel-item"
+            role="group"
+            aria-roledescription="slide"
+            aria-label="${x => x.getActiveIndex()} of ${x => x.getItemCount()}"
+        >
+            ${x => x.getActiveItem()}
+        </div>
+    </div>
+    <slot name="previous-flipper"> SEE DEFAULT </slot>
+    <slot name="tab-list">SEE DEFAULT </slot>
+    </div>
+</template>
+```
+
+**Default Flipper**
+```html
```

Comment:
Having default buttons poses an interesting question. These elements will need localized text labels to communicate to AT what they are for. Typically we don't bake localized strings into elements though.

Reply:
Good thoughts. I assume we want implementors/devs to be able to use their own flippers, so I'll add this concern to the list and see if I can come up with something.

Reply:
I think one way to approach this is how we default the control. For instance, the default carousel would like quite a bit like our current implementation with sequence indicators and flippers which are hidden from AT. This way, the default flippers are hidden from AT by design. To unhide them you would need to pass your own flippers. We could also easily ensure that buttons match the styling by using the slotted pseudo selector to assign the styles to any button passed - `::slotted(button)`. I think there's a lot of room to creatively solve this.

Reply:
Do we want buttons to have a certain styling, or do we want the author/implementor to pass all that in?
Maybe a sync on this idea and others is a good idea. I'll get that going soon

Reply:
@chrisdholt Does your fast-flipper address all of this? I took the defaults out for the spec and actually have the **Implementation** using fast-flipper.

Reply:
There's likely a couple ways to handle it:

1) Take a dependency on Flipper.

2) Support default buttons in the slot which use the fast-flipper styles. Doing this would allow reuse if someone wanted to pass the fast-flipper themselves and handle telemetry, etc.

3) Some other way?

Reply:
> @chrisdholt Does your fast-flipper address all of this? I took the defaults out for the spec and actually have the **Implementation** using fast-flipper.

Correction: I added fast-flipper as the default for the next-button and previous-button slots in the template.